tidy3d/components/data/data_array.py

```diff
@@ -352,15 +380,15 @@ def interp(
 
     def conj(self, *args: Any, **kwargs: Any):
         """Return the complex conjugate of this DataArray."""
-        if self.tracers is not None:
-            return self.__array_wrap__(anp.conj(self.tracers))
+        if isbox(self.values.flat[0]) or AUTOGRAD_KEY in self.attrs:
+            return self.insert_tracers(anp.conj(anp.array(self.values.tolist())))
         return super().conj(*args, **kwargs)
 
     @property
     def real(self):
         """Return the real part of this DataArray."""
-        if self.tracers is not None:
-            return self.__array_wrap__(anp.real(self.tracers))
+        if isbox(self.values.flat[0]) or AUTOGRAD_KEY in self.attrs:
```

Comment:
perhaps we can refactor this logic out (eg. the one in real, conj, etc).

```py
def apply_traced(self, fn):

    if isbox(self.values.flat[0]) or AUTOGRAD_KEY in self.attrs:
        tracers = anp.array(self.values.tolist()))
        new_values = fn(tracers)
        return self.insert_tracers(new_values)

    return fn(self)
    
@property
def real(self):
    return self.apply_traced(anp.real)
```

Reply:
also it seems the `anp.array(self.values.tolist())` and `if` statement both could use the convenience methods / functions used to check these things?

Reply:
Not really sure how to handle the `super()` call this way? But yeah I can at least partially refactor this..

Reply:
do we need super or can `np.real(self)` suffice?

Reply:
wouldn't that be infinite recursion? pretty sure `np.real(self)` will call `self.real`